### PR TITLE
Create functioning skeleton for VM Translator 

### PR DESCRIPTION
**Pull request by wildmaples**
### Changes

- Add a VM Translator with Parser and Code Writer hooked up. 
- Add a bin script for the VM Translator 

The Code Writer is not completed yet, but it does have basic functionality for `SimpleAdd.asm`: which includes push constant and arithmetic add.  

Additional work will be done for Code Writer in another PR.

### CODE REVIEW

**Review by tomstuart — APPROVED** [review]
Great start! Giving this a 👍🏻 because everything seems to be working so far. I’ve left a few comments with things to think about, but nothing that should prevent you from merging. 🎉

File: lib/code_writer.rb [hunk 1]
```diff
@@ -15,6 +15,15 @@ def write_push_pop(command, segment, index)
     EOF
   end
 
+  def write_arithmetic(command)
+    @out.puts <<~EOF
+      AM=M-1
+      D=M
+      A=A-1
+      M=M+D
```
[the remark this answers is not in the document]
Comment (wildmaples) [reply lib/code_writer.rb:23]: Currently `#write_arithmetic` is hard coded to add two numbers. If we were to break down the assembly code into modular chunks, I think we need to separate each piece of action. For example, two pushes and then the logic for the add.

The assumptions we are making in the current code is that 1) there are two parameters (not the case for `not` or `neg`), and 2) the arithmetic computation can be done in a single line of assembly code (not the case for other commands). So... the current implementation isn't reliable yet. 

When we start supporting other arithmetic commands, we'll have to break apart this method into more modular parts so it's easier to return the correct assembly code.

[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/code_writer.rb:23]: That’s not quite what I was getting at. You’re certainly correct that we’ve assumed two arguments which won’t be the case for all arithmetic operations, but on the “separating each piece” point, `#write_arithmetic` is already only generating the code for the addition itself — the pushes are done separately by `#write_push_pop`.

Anyway, it’s not important as long as the code works, which it does at the moment. We can come back to this topic if anything goes wrong in future!

File: test/integration/vm_translator_test.rb [hunk 2]
```diff
@@ -15,4 +15,30 @@ def test_integration_test
...
+      @SP 
+      M=M+1
+      AM=M-1
+      D=M
+      A=A-1
+      M=M+D
+    EOF
+
+    assert_equal(expected, assembly_code)
+  end
```
[the remark this answers is not in the document]
Comment (tomstuart) [reply test/integration/vm_translator_test.rb:43]: I’ve opened #4 to illustrate how this might work.

**Pull request merged**